lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -84,24 +84,44 @@ def add_node_data(node)
   end
 
   # Rewrite a given file resource with the metadata from a fileserver based file
-  def replace_metadata(resource, metadata, recurse = false)
+  def replace_metadata(resource, metadata)
+    # TODO: when we stop using to_ral, we'll have to add a case for ensure => <path>
+    # implying a link.
     if resource[:links] == "manage" && metadata.ftype == "link"
       resource.delete(:source)
-      resource[:ensure] = "link"
       resource[:target] = metadata.destination
-    elsif resource[:links] == "follow"
-      resource[:ensure] = metadata.ftype
     end
 
+    # Most of the time, ensure should match the returned metadata. Exceptions are when
+    # ensure is `present` (it could be anything), or a path (implies creating a link).
+    # `to_ral` caused the path case to munge to managing links, so we warn if there's
+    # a mismatch that's not `present`.
+    if resource[:ensure] != :present && resource[:ensure] != metadata.ftype
+      Puppet.warning "#{resource} expected to create a #{resource[:ensure]}, but its source "+
+        "is a #{metadata.ftype}. The source metadata will override the 'ensure' property."
+    end
+
+    resource[:ensure] = metadata.ftype
+
     if metadata.ftype == "file"
+      # Since we use the checksum from metadata, it makes sense to pick up the checksum type
+      # from the same source. This should never differ, but seems like the most correct
+      # expression of intent.
       resource[:checksum_value] = sumdata(metadata.checksum)
+      resource[:checksum] = metadata.checksum_type
     end
 
-    if recurse
-      resource[:source] = metadata.source
-      if metadata.ftype == "file"
-        resource[:checksum] = metadata.checksum_type
+    case resource[:source_permissions]
+    when "use"
+      resource[:owner] = metadata.owner if (resource[:owner].nil? || resource[:owner] == :absent)
+      resource[:group] = metadata.group if (resource[:group].nil? || resource[:group] == :absent)
+      if resource[:mode].nil? || resource[:mode] == :absent
```

Comment:
I took that from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file/source.rb#L252, but since we're operating on a Resource it would never be `:absent`. I'll remove that check.